File: buscas_locais/simulated_annealing.py
import math
import random
import time
from typing import TYPE_CHECKING
from buscas_locais.busca_local import BuscaLocal

if TYPE_CHECKING:
    from labirinto import LabirintoBusca

class SimulatedAnnealing(BuscaLocal):
    
    def __init__(
            self,
            labirinto: "LabirintoBusca",
            temperatura_inicial: float = 100.0,
            temperatura_final: float = 0.01,
            fator_resfriamento: float = 0.85):

        super().__init__(labirinto)
        self.temperatura_inicial = temperatura_inicial
        self.temperatura_final = temperatura_final
        self.fator_resfriamento = fator_resfriamento

    def executar(self):
        temperatura = self.temperatura_inicial

        atual = self.gerar_solucao_inicial()
        custo_atual, caminho_atual = self.custo(atual)
        custo_inicial = custo_atual
        
        melhor_solucao = atual.copy()
        melhor_custo = custo_atual
        melhor_caminho = caminho_atual

        historico = [custo_atual]
        iteracoes = 0

        inicio_tempo = time.time()

        while temperatura > self.temperatura_final:
            
            vizinho = self.gerar_vizinho(atual)
            custo_vizinho, caminho_vizinhos = self.custo(vizinho)

            delta = custo_vizinho - custo_atual

            ########################################################
            # ACEITA SOLUÇÕES MELHORES
            ########################################################
            if delta < 0:
                atual = vizinho
                custo_atual = custo_vizinho
                caminho_atual = caminho_vizinhos

            ########################################################
            # ACEITA SOLUÇÕES PIORES COM CERTA PROBABILIDADE
            ########################################################
            else:
                probabilidade = math.exp(-delta / temperatura)
                
                if random.random() < probabilidade:
                    atual = vizinho
                    custo_atual = custo_vizinho
                    caminho_atual = caminho_vizinhos

            ########################################################
            if custo_atual < melhor_custo:
                melhor_solucao = atual.copy()
                melhor_custo = custo_atual
                melhor_caminho = caminho_atual

            historico.append(custo_atual)
            temperatura *= self.fator_resfriamento
            iteracoes += 1

        tempo_execucao = time.time() - inicio_tempo

        return {
            "melhor_solucao": melhor_solucao,
            "caminho": melhor_caminho,
            "melhor_custo": melhor_custo,
            "media_custo": sum(historico) / len(historico),
            "pior_custo": max(historico),
            "historico": historico,
            "iteracoes": iteracoes,
            "tempo_execucao": tempo_execucao,
            "is_taxa_aceitavel": custo_atual < custo_inicial * 1.1,
            "houve_melhora": melhor_custo < custo_inicial
        }

File: interface-web/scripts/buscasLocais/simulatedAnnealing.js
import { BuscaLocal } from './buscaLocal.js';

export class SimulatedAnnealing extends BuscaLocal {
  constructor(
    labirinto,
    temperaturaInicial = 100.0,
    temperaturaFinal = 0.01,
    fatorResfriamento = 0.85
  ) {
    // Repassa o labirinto para a classe pai (BuscaLocal)
    super(labirinto);
    
    // Configurações específicas do SA
    this.temperaturaInicial = temperaturaInicial;
    this.temperaturaFinal = temperaturaFinal;
    this.fatorResfriamento = fatorResfriamento;
  }

  executar() {
    // Usa os métodos herdados da classe base
    let atual = this.gerarSolucaoInicial();
    let [custoAtual, caminhoAtual] = this.custo(atual);
    const custoInicial = custoAtual;

    let melhorSolucao = [...atual];
    let melhorCusto = custoAtual;
    let melhorCaminho = [...caminhoAtual];

    const historico = [custoAtual];
    let iteracoes = 0;
    let temperatura = this.temperaturaInicial;

    const inicioTempo = performance.now();

    // Loop principal até que a "temperatura" esfrie
    while (temperatura > this.temperaturaFinal) {
      iteracoes++;

      // 1. Gera um ÚNICO vizinho aleatório por iteração (herdado)
      const vizinho = this.gerarVizinho(atual);
      const [custoVizinho, caminhoVizinho] = this.custo(vizinho);

      // Delta: diferença de custo (como queremos minimizar, delta < 0 é melhor)
      const delta = custoVizinho - custoAtual;

      // 2. Aceita automaticamente se a nova solução for melhor (delta negativo)
      if (delta < 0) {
        atual = vizinho;
        custoAtual = custoVizinho;
        caminhoAtual = caminhoVizinho;
      } 
      // 3. Se for pior, aceita baseando-se na probabilidade térmica
      else {
        const probabilidade = Math.exp(-delta / temperatura);
        
        // Sorteia um número entre 0 e 1 e compara com a probabilidade
        if (Math.random() < probabilidade) {
          atual = vizinho;
          custoAtual = custoVizinho;
          caminhoAtual = caminhoVizinho;
        }
      }

      // Atualiza a melhor solução global encontrada até agora (para não perdê-la)
      if (custoAtual < melhorCusto) {
        melhorSolucao = [...atual];
        melhorCusto = custoAtual;
        melhorCaminho = [...caminhoAtual];
      }

      historico.push(custoAtual);
      
      // 4. Resfria o sistema
      temperatura *= this.fatorResfriamento;
    }

    const tempoExecucao = (performance.now() - inicioTempo) / 1000;

    return {
      melhorSolucao: melhorSolucao,
      caminho: melhorCaminho,
      custoFinal: custoAtual,
      custoInicial: custoInicial,
      melhorCusto: melhorCusto,
      mediaCusto: historico.reduce((a, b) => a + b, 0) / historico.length,
      piorCusto: Math.max(...historico),
      historico: historico,
      iteracoes: iteracoes,
      tempoExecucao: tempoExecucao,
      isTaxaAceitavel: melhorCusto <= custoInicial,
      houveMelhora: melhorCusto < custoInicial
    };
  }
}